src/communication/channelmanager.cpp

```diff
@@ -0,0 +1,280 @@
+        return err;
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error ChannelManager::Close()
+{
+    {
+        UniqueLock lock(mMutex);
+
+        LOG_DBG() << "Close channel manager";
+
+        mClose = true;
+        mCondVar.NotifyAll();
+
+        if (!mTransport->IsOpened()) {
+            return ErrorEnum::eNone;
+        }
+
+        if (auto err = mTransport->Close(); !err.IsNone()) {
+            return err;
+        }
+    }
+
+    return mThread.Join();
+}
+
+int ChannelManager::Write(uint32_t port, const void* data, size_t size)
+{
+    LOG_DBG() << "Write channel: port=" << port << " size=" << size;
+
+    auto header = PrepareHeader(port, aos::Array<uint8_t>(reinterpret_cast<const uint8_t*>(data), size));
+    if (!header.mError.IsNone()) {
+        LOG_ERR() << "Failed to prepare header: error=" << header.mError.Message();
+
+        return -EINVAL;
+    }
+
+    aos::UniqueLock lock(sWriteMutex);
+
+    if (auto ret = mTransport->Write(&header.mValue, sizeof(AosProtocolHeader));
+        ret < static_cast<int>(sizeof(AosProtocolHeader))) {
+        return ret;
+    }
+
+    return mTransport->Write(data, size);
+}
+
+bool ChannelManager::IsConnected() const
+{
+    return mTransport->IsOpened();
+}
+
+Error ChannelManager::Run()
+{
+    return mThread.Run([this](void*) {
+        while (true) {
+            {
+                aos::UniqueLock lock(mMutex);
+                if (mClose) {
+                    return aos::ErrorEnum::eNone;
+                }
+            }
+
+            if (auto err = TryConnect(); !err.IsNone()) {
+                LOG_ERR() << "Failed to connect: error=" << err.Message();
+
+                continue;
+            }
+
+            if (auto err = HandleRead(); !err.IsNone()) {
+                LOG_ERR() << "Failed to handle read: error=" << err.Message();
+            }
+
+            CloseChannels();
+
+            {
+                aos::UniqueLock lock(mMutex);
+                if (mClose) {
+                    return aos::ErrorEnum::eNone;
+                }
+            }
+
+            sleep(cReconnectPeriod);
+        }
+    });
+}
+
+Error ChannelManager::HandleRead()
+{
+    while (true) {
+        {
+            aos::UniqueLock lock(mMutex);
+
+            mCondVar.Wait(lock, [this] { return !mChannels.IsEmpty() || mClose; });
+
+            if (mClose) {
+                return ErrorEnum::eNone;
+            }
+        }
+
+        if (auto ret = mTransport->Read(mHeader.Get(), sizeof(AosProtocolHeader));
+            ret < static_cast<int>(sizeof(AosProtocolHeader))) {
+            return {ErrorEnum::eRuntime, "Failed to read header"};
+        }
+
+        AosProtocolHeader header;
+        memcpy(&header, mHeader.Get(), sizeof(AosProtocolHeader));
+
+        uint32_t port    = header.mPort;
+        uint32_t msgSize = header.mDataSize;
+
+        UniqueLock lock(mMutex);
+
+        for (auto& channel : mChannels) {
+            if (channel.mPort == port) {
```

Comment:
just find channel by port
case when channel not found is not handled

Reply:
Done